service/csvService.go

```diff
@@ -6,133 +6,219 @@ import (
+}
 
-// }
-func getAndSaveData() {
+func filterRecord(s []*structs.Field, m map[string]string) bool {
+	for _, f := range s {
+		log.Info("validating "+f.Name()+" in conditions "+m[f.Name()])
+		if filter := m[f.Name()]; filter != "" {
+			if f.Value().(string) != filter {
+				return false
+			}
+		}
 
-	response, err := http.Get("http://openlibrary.org/search.json?q=the+lord+of+the+rings&page=1")
+	}
+	return true
+}
 
+func deleteCsv() {
+	err := os.Remove(getConfigService().GetConfig().CSV.FileName)
 	if err != nil {
-		fmt.Println(err.Error())
-		panic(err)
+		log.Error("Cannot delete file", err)
 	}
+}
 
-	responseData, err := ioutil.ReadAll(response.Body)
+func getAndSaveData(topic string) error {
+	// the+lord+of+the+rings&page=1
+	// <search-term>  <page>
+
+	deleteCsv() // before write results clears file
+	escapedTopic := strings.ReplaceAll(topic, 	" ", "+")
+	apiUrl := strings.ReplaceAll(getConfigService().GetConfig().External.ApiUrl, "<topic>", escapedTopic)
+	file, err := openOrCreate(getConfigService().GetConfig().CSV.FileName)
 
 	if err != nil {
-		log.Fatal(err)
+		log.Error("Unable to use file", err)
+		return err
 	}
 
-	var responseObject SearchResponse
+	for i := 1; true; i++ {
+		response, err := http.Get(strings.ReplaceAll(apiUrl, "<page>", strconv.Itoa(i)))
+
+		if err != nil {
+			log.Info(err.Error())
+			return err
+		}
 
-	json.Unmarshal(responseData, &responseObject)
 
-	file, err := os.Create(CSVFile)
-	defer file.Close()
+		responseData, err := ioutil.ReadAll(response.Body)
 
-	if err != nil {
-		log.Fatal("Cannot create file", err)
+		if err != nil {
+			log.Error(err)
+			return err
+		}
+
+		var responseObject model.SearchResponse
+
+		json.Unmarshal(responseData, &responseObject)
+
+		if (len(responseObject.Doc) == 0) {
+			break
+		}
+
+		saveCsv(responseObject, *file)
 	}
-	SaveCsv(responseObject, *file)
 
+	defer file.Close()
+
+	return nil
 }
 
-func SaveCsv(responseObject SearchResponse, file os.File) {
+func saveCsv(responseObject model.SearchResponse, file os.File) {
 	writer := csv.NewWriter(&file)
-	defer writer.Flush()
 	for _, doc := range responseObject.Doc {
-		err := writer.Write(interfaceToString(structs.Values(doc)))
+		str := interfaceToString(structs.Values(doc))
+		err := writer.Write(str)
 		if err != nil {
-			log.Fatal("cannot write CSV", err)
+			log.Error("cannot write CSV", err)
 		}
 	}
+	defer writer.Flush()
+}
+func openOrCreate(path string) (*os.File, error) {
+	file, _ := os.Open(path)
+
+	if file != nil {
+		log.Info("File opened")
+		return file, nil
+	}
+
+	file, err := os.Create(path)
+
+	if err != nil {
+		log.Error("File was not able to create")
+		return nil, err
+	}
+	return file, nil
 }
 
 func interfaceToString(record []interface{}) []string {
-	var a []string
+	a := make([]string, len(record))
 
-	for _, row := range record {
-		a = append(a, row.(string))
+	for i, row := range record {
+		a[i] = row.(string)
 	}
 	return a
 }
+
+func getConfigService() ConfigService {
```

Comment:
What is ConfigService? I can't find it. _(Not really sure why/how your project runs BTW)_

Reply:
Config service is just right here: https://github.com/oscarSantoyo/academy-go-q12021/pull/1/files#diff-30d4dba7eb0fea6bedd04886a16e6d9c6b00758616d9afe00474295e9039fedb

Reply:
Here are the instructions on how to run: https://github.com/oscarSantoyo/academy-go-q12021/pull/1/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R16

Reply:
Not sure what you mean by 'not really sure why your project runs', please elaborate your queries so I can help you out.

Basically the center point here is the dependency injection and the singletons container, based on the services interfaces is decoupling services from where you use them and you ask for a service using its interface. This is basically what the article that talks about their implementation of clan architecture wants to achieve, but my approach is less complex and less redundant on the responsibility chain.

The container is responsible of providing the service you require, so you don't have to create it.